torch_geometric/utils/negative_sampling.py

```diff
@@ -99,23 +104,34 @@ def structured_negative_sampling(edge_index, num_nodes=None):
         edge_index (LongTensor): The edge indices.
         num_nodes (int, optional): The number of nodes, *i.e.*
             :obj:`max_val + 1` of :attr:`edge_index`. (default: :obj:`None`)
+        contains_neg_self_loops (bool, optional): If set to
+            :obj:`False`, sampled negative edges will not contain self loops.
+            It is independent of :attr:`edge_index`. (default: :obj:`True`)
 
     :rtype: (LongTensor, LongTensor, LongTensor)
     """
     num_nodes = maybe_num_nodes(edge_index, num_nodes)
 
+    assert is_neg_samp_feasible(edge_index, 'structure', num_nodes, False,
+                                contains_neg_self_loops)
+
+    self_loops = torch.arange(num_nodes) * (num_nodes + 1)
     i, j = edge_index.to('cpu')
     idx_1 = i * num_nodes + j
 
-    k = torch.randint(num_nodes, (i.size(0), ), dtype=torch.long)
+    k = torch.randint(num_nodes, (i.size(0),), dtype=torch.long)
     idx_2 = i * num_nodes + k
 
     mask = torch.from_numpy(np.isin(idx_2, idx_1)).to(torch.bool)
```

Comment:
I think we need the to modify the `mask`  before the while loop too. Because lets say none of the nodes are true negatives then rest would be empty and the check for self loops wouldn't run.
```suggestion
    mask = torch.from_numpy(np.isin(idx_2, idx_1)).to(torch.bool)
    if not contains_neg_self_loops:
        self_loop_mask = torch.from_numpy(np.isin(idx_2, self_loops)).to(
                torch.bool)
        mask = torch.logical_or(mask, self_loop_mask)
```